clippy_lints/src/multiple_unsafe_ops_per_block.rs

```diff
@@ -100,18 +102,41 @@ impl<'tcx> LateLintPass<'tcx> for MultipleUnsafeOpsPerBlock {
 struct UnsafeExprCollector<'tcx> {
     tcx: TyCtxt<'tcx>,
     typeck_results: &'tcx TypeckResults<'tcx>,
-    unsafe_ops: Vec<(&'static str, Span)>,
+    unsafe_ops: FxHashMap<Span, &'static str>,
 }
 
 impl<'tcx> UnsafeExprCollector<'tcx> {
     fn collect_unsafe_exprs(cx: &LateContext<'tcx>, block: &'tcx hir::Block<'tcx>) -> Vec<(&'static str, Span)> {
         let mut collector = Self {
             tcx: cx.tcx,
             typeck_results: cx.typeck_results(),
-            unsafe_ops: vec![],
+            unsafe_ops: FxHashMap::default(),
         };
         collector.visit_block(block);
-        collector.unsafe_ops
+        #[allow(
+            rustc::potential_query_instability,
+            reason = "span ordering only needed inside the one expression being walked"
+        )]
+        let mut unsafe_ops = collector
+            .unsafe_ops
+            .into_iter()
+            .map(|(span, msg)| (msg, span))
+            .collect::<Vec<_>>();
+        unsafe_ops.sort_unstable();
+        unsafe_ops
+    }
+}
+
+impl UnsafeExprCollector<'_> {
+    fn insert_span(&mut self, span: Span, message: &'static str) {
+        if span.from_expansion() {
+            self.unsafe_ops.insert(
+                span.source_callsite(),
```

Comment:
You still need to walk this up to the unsafe block's context to handle nested macros properly. e.g.
```rust
macro_rules! m {
  () => {
    unsafe_call();
    unsafe_mac!();
  }
}
```